### PR TITLE
Allow skipping ranges of versions

### PR DESCRIPTION
**Pull request by pgomulka**
Multiple version ranges are allowed to be used in section skip in yml
tests. This is useful when a bugfix was backported to latest versions
and all previous releases contain a wire breaking bug.
examples:
`6.1.0 - 6.3.0, 6.6.0 - 6.7.9, 7.0 -`
`- 7.2, 8.0.0 -`

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java [hunk 1]
```diff
@@ -98,33 +99,30 @@ public static SkipSection parse(XContentParser parser) throws IOException {
...
         this.reason = null;
     }
 
     public SkipSection(String versionRange, List<String> features, String reason) {
         assert features != null;
-        Version[] versions = parseVersionRange(versionRange);
-        this.lowerVersion = versions[0];
-        this.upperVersion = versions[1];
+        this.versionRanges = parseVersionRanges(versionRange);
+        assert versionRanges.size() >= 1;
```
Comment (pugnascotia) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java]: Would this be clearer?

```suggestion
        assert versionRanges.isEmpty() == false;
```

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java [hunk 2]
```diff
@@ -139,8 +137,11 @@ public boolean skip(Version currentVersion) {
         if (isEmpty()) {
             return false;
         }
-        boolean skip = lowerVersion != null && upperVersion != null && currentVersion.onOrAfter(lowerVersion)
-            && currentVersion.onOrBefore(upperVersion);
+        boolean skip = false;
+        for (VersionRange versionRange : versionRanges) {
+            skip |= versionRange.contain(currentVersion);
```
Comment (pugnascotia) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java]: What about:

```java
boolean skip = versionRanges.stream().anyMatch(range -> range.contain(currentVersion));

return skip || Features.areAllSupported(features) == false;
```

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java [hunk 3]
```diff
@@ -153,24 +154,34 @@ public boolean isEmpty() {
...
+        if (rawRanges.trim().equals("all")) {
+            return Collections.singletonList(new VersionRange(VersionUtils.getFirstVersion(), Version.CURRENT));
         }
-        String[] skipVersions = versionRange.split("-");
-        if (skipVersions.length > 2) {
-            throw new IllegalArgumentException("version range malformed: " + versionRange);
+        String[] ranges = rawRanges.split(",");
+        List<VersionRange> versionRanges = new ArrayList<>();
+        for (String rawRange : ranges) {
+            String[] skipVersions = wrapWithSpaces(rawRange).split("-");
```
Comment (pugnascotia) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java]: I don't understand why `wrapWithSpaces` is necessary here?

Comment (pugnascotia) [reply test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java]: Ah, it's to ensure you get two elements in the `skipVersions` array. That's a little subtle.

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    public Version lower() {
+        return lower;
+    }
+
+    public Version upper() {
+        return upper;
+    }
+
+    public boolean contain(Version currentVersion) {
```
Comment (pugnascotia) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java]: ...I don't suppose you'd rename this to `contains`? It's just that I'm so used to `Collection.contains` 😅

Comment (pgomulka) [reply test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java]: it is a good idea, will do

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java (hunk 3 above)
Comment (pgomulka) [reply test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java]: yes, definitely not needed.

Comment (pgomulka) [reply test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java]: ok so I see why I put that there. I wanted to allow formats like
`8.0.0 -` where there is no additional space. Also didn't want to make it a special case.
how about using `"8.0.0 -".split("-",-1)` ? -1 there means that the length of array element can have any element. With 0 it means that trailing empty element will not be returned

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+public class VersionRange {
+    private final Version lower;
+    private final Version upper;
+
+    public VersionRange(Version lower, Version upper) {
+        this.lower = lower;
+        this.upper = upper;
+    }
+
+    public Version lower() {
```
Comment (pugnascotia) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java]: Do we not usually use bean-style naming, i.e. `getLower` and `getUpper`?

Comment (pgomulka) [reply test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/VersionRange.java]: you are right, will fix

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/180_locale_dependent_mapping.yml [hunk 6]
```diff
@@ -1,8 +1,5 @@
 ---
 "Test Index and Search locale dependent mappings / dates":
-  - skip:
-      version: " - 6.1.99"
```
Comment (pugnascotia) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search/180_locale_dependent_mapping.yml]: Is this redundant now?

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java [hunk 7]
```diff
@@ -139,8 +137,7 @@ public boolean skip(Version currentVersion) {
         if (isEmpty()) {
             return false;
         }
-        boolean skip = lowerVersion != null && upperVersion != null && currentVersion.onOrAfter(lowerVersion)
-            && currentVersion.onOrBefore(upperVersion);
+        boolean skip = versionRanges.stream().anyMatch(range -> range.contains(currentVersion));
         skip |= Features.areAllSupported(features) == false;
```
Comment (pugnascotia) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java]: Do you think it's worth changing this to the following?

```java
return skip || Features.areAllSupported(features) == false;
```

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java [hunk 8]
```diff
@@ -153,24 +150,34 @@ public boolean isEmpty() {
...
+        return versionRanges;
+    }
 
-        String lower = skipVersions[0].trim();
-        String upper = skipVersions[1].trim();
-        return new Version[] {
-            lower.isEmpty() ? VersionUtils.getFirstVersion() : Version.fromString(lower),
-            upper.isEmpty() ? Version.CURRENT : Version.fromString(upper)
-        };
+    private static String wrapWithSpaces(String rawRange) {
```
Comment (pugnascotia) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/section/SkipSection.java]: Can we remove this now?

**Review by pugnascotia — APPROVED** [review]

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/180_locale_dependent_mapping.yml (hunk 6 above)
Comment (pgomulka) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/180_locale_dependent_mapping.yml]: so in master I think it does not make sense. The ` - 6.1.99` is translated into `7.0.0 - 6.1.99` And master runs bwc tests from 7.0 - 8 range

**Pull request merged**